File: server/src/utils/retry.ts
/**
 * Exponential backoff retry wrapper for async operations.
 */
export async function retry<T>(
  fn: () => Promise<T>,
  maxAttempts = 3,
  baseDelayMs = 1000
): Promise<T> {
  let lastError: Error | unknown;

  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (err: unknown) {
      lastError = err;
      
      const status = (err as any)?.response?.status;
      const isClientError = status && (status === 400 || status === 401 || status === 404);
      
      const isRateLimit =
        err instanceof Error &&
        (err.message.includes('429') ||
          err.message.toLowerCase().includes('rate limit') ||
          err.message.toLowerCase().includes('quota') ||
          status === 403);

      if (isClientError && !isRateLimit) {
        console.warn(`Client error ${status} encountered, aborting retries:`, (err as Error).message);
        break;
      }

      if (attempt < maxAttempts && isRateLimit) {
        const delay = baseDelayMs * Math.pow(2, attempt - 1);
        console.warn(`Rate limit hit. Retrying in ${delay}ms (attempt ${attempt}/${maxAttempts})`);
        await sleep(delay);
      } else if (attempt < maxAttempts) {
        // Retry other errors too with smaller delay
        const delay = baseDelayMs * attempt;
        console.warn(`Request failed. Retrying in ${delay}ms (attempt ${attempt}/${maxAttempts}):`, (err as Error).message);
        await sleep(delay);
      } else {
        break;
      }
    }
  }

  throw lastError;
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: server-python/app/utils/retry.py
import asyncio
from typing import TypeVar, Callable, Any

T = TypeVar("T")


async def retry(
    fn: Callable[..., T],
    max_attempts: int = 3,
    base_delay_ms: int = 1000,
    *args: Any,
    **kwargs: Any,
) -> T:
    """Exponential backoff retry wrapper for async operations."""
    last_error = None

    for attempt in range(1, max_attempts + 1):
        try:
            if asyncio.iscoroutinefunction(fn):
                return await fn(*args, **kwargs)
            else:
                return fn(*args, **kwargs)
        except Exception as err:
            last_error = err

            # Safely check for client errors (don't retry)
            status = getattr(err, "status", None)
            if not status and hasattr(err, "response"):
                try:
                    resp = getattr(err, "response")
                    if hasattr(resp, "get"):
                        status = resp.get("status")
                    elif hasattr(resp, "status_code"):
                        status = resp.status_code
                    elif hasattr(resp, "status"):
                        status = resp.status
                except Exception:
                    pass
            is_client_error = status and status in [400, 401, 404]

            # Check for rate limits
            err_msg = str(err).lower()
            is_rate_limit = (
                "429" in err_msg
                or "rate limit" in err_msg
                or "quota" in err_msg
                or status == 403
            )

            if is_client_error and not is_rate_limit:
                print(f"Client error {status} encountered, aborting retries: {err}")
                break

            if attempt < max_attempts:
                if is_rate_limit:
                    delay = base_delay_ms * (2 ** (attempt - 1))
                    print(
                        f"Rate limit hit. Retrying in {delay}ms (attempt {attempt}/{max_attempts})"
                    )
                else:
                    delay = base_delay_ms * attempt
                    print(
                        f"Request failed. Retrying in {delay}ms (attempt {attempt}/{max_attempts}): {err}"
                    )
                await asyncio.sleep(delay / 1000)
            else:
                break

    raise last_error


def sleep(ms: int) -> asyncio.sleep:
    """Sleep for the specified milliseconds."""
    return asyncio.sleep(ms / 1000)
